### PR TITLE
Implemented triggerOffset property.

### PR DESCRIPTION
**Pull request by mariuszzak**
I have implemented triggerOffset property which let you adjust starting loading more items before the scroll is on the very bottom of window (or scrollable div). I've also resolved problem described here: https://github.com/hhff/ember-infinity/issues/82. Now the trigger in scrollable divs works properly :)

### CODE REVIEW

**Comment by mkorfmann** [comment]
Really nice, you're the man!

Though it looks like the build is failing, do you know what's the issue there?

CheeriO

**Comment by mariuszzak** [comment]
Thanks! :) 
I'm trying to find the problem. On my local environment tests pass properly.

**Comment by mkorfmann** [comment]
Hehe, good ol "works for me" :)

**Comment by hhff** [comment]
I would like to explore using https://github.com/dockyard/ember-in-viewport for viewport checking - it has offsets & tolerance and everything we need.  What do u think @mariuszzak ?

File: addon/components/infinity-loader.js [hunk 1]
```diff
@@ -38,14 +39,34 @@ const InfinityLoaderComponent = Ember.Component.extend({
...
-    var scrollableBottom = scrollable.height() + scrollable.scrollTop();
+  _selfOffset() {
+    var selfOffset;
+    if(this.get('_customScrollableIsDefined')) {
+      selfOffset = this.$().position().top + this.get("_scrollable").scrollTop();
+    }
+    else {
+      selfOffset = this.$().offset().top;
+    }
+    return selfOffset;
```
Comment (mike-north) [inline addon/components/infinity-loader.js]: more of a stylistic point, but you could write this as 

```js
if(this.get('_customScrollableIsDefined')) {
  return this.$().position().top + this.get("_scrollable").scrollTop();
}
else {
  return this.$().offset().top;
}
```
or even
```js
return this.get('_customScrollableIsDefined')) ?
  this.$().position().top + this.get("_scrollable").scrollTop() :
  this.$().offset().top;
```

File: tests/acceptance/infinity-route-offset-trigger-test.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  andThen(() => {
+    var postList = find('ul');
+    var infinityLoader = find('.infinity-loader');
+
+    assert.equal(postList.find('li').length, 25, "Two items should be in the list");
+    assert.equal(infinityLoader.hasClass('reached-infinity'), false, "Infinity should not yet have been reached");
+    var triggerOffset = postList.get(0).scrollHeight - postList.height();
+    postList.scrollTop(triggerOffset - 20);
+    triggerEvent('ul', 'scroll');
+    andThen(() => {
```
Comment (mike-north) [inline tests/acceptance/infinity-route-offset-trigger-test.js]: `andThen`s should be all at the top level closure of the test

```js

andThen(() => { ... });
andThen(() => { ... });
click('#myId');
andThen(() => { ... });
```

File: tests/acceptance/infinity-route-offset-trigger-test.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+  andThen(() => {
+    var postList = find('ul');
+    var infinityLoader = find('.infinity-loader');
+
+    assert.equal(postList.find('li').length, 25, "Two items should be in the list");
+    assert.equal(infinityLoader.hasClass('reached-infinity'), false, "Infinity should not yet have been reached");
+    var triggerOffset = postList.get(0).scrollHeight - postList.height();
+    postList.scrollTop(triggerOffset - 20);
+    triggerEvent('ul', 'scroll');
```
Comment (mike-north) [inline tests/acceptance/infinity-route-offset-trigger-test.js]: `triggerEvent` is an async test helper, just like `andThen` and `click`. It should be in the top level closure

see: http://guides.emberjs.com/v1.10.0/testing/test-helpers/#toc_asynchronous-helpers
```js
andThen(...);
triggerEvent(...);
click(...);
```

File: tests/acceptance/infinity-route-offset-trigger-test.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  teardown() {
+    Ember.run(App, 'destroy');
+    server.shutdown();
+  }
+});
+
+
+test('it should start loading more items when the scroll is on the very bottom ' +
+  'when triggerOffset is not set', assert => {
+  assert.expect(8);
```
Comment (mike-north) [inline tests/acceptance/infinity-route-offset-trigger-test.js]: Once you straighten out the async test helper issue, you won't need assert.expect at all anymore

File: tests/dummy/app/routes/demo-scrollable.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      var fd = fakeData;
+      var page = parseInt(request.queryParams.page, 10);
+      var per =  parseInt(request.queryParams.per_page, 10);
+      var payload = {
+        posts: fd.slice((page - 1) * per, Math.min((page) * per, fd.length)),
+        meta: {
+          total_pages: Math.ceil(fd.length/per)
+        }
+      };
+      return [200, {}, JSON.stringify(payload)];
```
Comment (mike-north) [inline tests/dummy/app/routes/demo-scrollable.js]: we might want to consider copying @stefanpenner's json helper for this

https://github.com/stefanpenner/ember-jobs/blob/master/tests/helpers/json.js

**Comment by hhff** [comment]
ur the man mike!  Happy to put the move to ember-in-viewport as work for after we get this in!

**Comment by mariuszzak** [comment]
I apologize for the long absence of answers.

@hhff I think it's not worth to couple this plugin with `ember-in-viewport` just for so tiny feature. I'm not enthusiast of running a combine-harvester for one ear of grain :)

@mike-north Thanks a lot for code review! Fixes are ready. Is there any possibility to get rid of duplication of code?

```javascript
    postList = find('ul');
    infinityLoader = find('.infinity-loader');
    triggerOffset = postList.get(0).scrollHeight - postList.height();
```

File: tests/dummy/app/templates/demo-scrollable.hbs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    <li>{{item.id}}. {{item.name}}</li>
+  {{/each}}
+  {{infinity-loader
+    infinityModel=content
+    loadingText="Loading more awesome records..."
+    loadedText="Loaded all the records!"
+    triggerOffset=500
+    scrollable=".demo-items"
+  }}
+</ul>
```
Comment (davidgoli) [inline tests/dummy/app/templates/demo-scrollable.hbs]: please observe proper HTML indentation

File: tests/dummy/app/templates/demo-scrollable.hbs [hunk 7]
```diff
@@ -0,0 +1,19 @@
+<div class="demo-view">
+
+<div class="fixed-to-top jumbo-header text-center">
+<a href="https://github.com/hhff/ember-infinity"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://camo.githubusercontent.com/52760788cde945287fbb584134c4cbc2bc36f904/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f77686974655f6666666666662e706e67" alt="Fork me on GitHub" data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_white_ffffff.png"></a>
```
Comment (davidgoli) [inline tests/dummy/app/templates/demo-scrollable.hbs]: let's wrap long lines

File: tests/acceptance/infinity-route-offset-trigger-test.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  });
+  triggerEvent('ul', 'scroll');
+  andThen(() => {
+    postList = find('ul');
+    infinityLoader = find('.infinity-loader');
+
+    assert.equal(postList.find('li').length, 50, "50 items should be in the list");
+    assert.equal(infinityLoader.hasClass('reached-infinity'), true, "Infinity should have been reached");
+  });
+});
```
Comment (davidgoli) [inline tests/acceptance/infinity-route-offset-trigger-test.js]: there's a lot of duplication between these 2 `test`s, please refactor that out

**Comment by mariuszzak** [comment]
The same error is on master branch currently. New release of ember-beta brought some changes. Any idea?

**Comment by hhff** [comment]
Looks like we're going to have to move the infinity-loader unit tests to integration tests

**Comment by mariuszzak** [comment]
Tests seem to be ok now. Is there anything I should refine?

**Comment by hhff** [comment]
:+1: LGTM - @davidgoli / @kellyselden wanna give this a final gaze?

File: addon/components/infinity-loader.js [hunk 9]
```diff
@@ -38,14 +39,32 @@ const InfinityLoaderComponent = Ember.Component.extend({
...
-    var inView = selfOffset < scrollableBottom;
+  _triggerOffset() {
+    return this._selfOffset() - this.get('triggerOffset');
+  },
+
+  _shouldLoadMore() {
+    if(this.get('developmentMode')) {
+      return false;
+    }
+    return this._bottomOfScrollableOffset() > this._triggerOffset();
```
Comment (davidgoli) [inline addon/components/infinity-loader.js]: I'd just reformat this to use more whitespace, like so:

```js
   if (this.get('developmentMode')) {
      return false;
    }

    return this._bottomOfScrollableOffset() > this._triggerOffset();
```

File: addon/components/infinity-loader.js [hunk 10]
```diff
@@ -38,14 +39,32 @@ const InfinityLoaderComponent = Ember.Component.extend({
...
 
-  _checkIfInView() {
-    var selfOffset       = this.$().offset().top;
-    var scrollable       = this.get("_scrollable");
-    var scrollableBottom = scrollable.height() + scrollable.scrollTop();
+  _selfOffset() {
+    if(this.get('_customScrollableIsDefined')) {
+      return this.$().position().top + this.get("_scrollable").scrollTop();
+    }
+    else {
```
Comment (davidgoli) [inline addon/components/infinity-loader.js]: use `} else {` for consistency

File: addon/components/infinity-loader.js [hunk 11]
```diff
@@ -38,14 +39,32 @@ const InfinityLoaderComponent = Ember.Component.extend({
     this.get('_scrollable').off(`${eventName}.${this.get('guid')}`);
   },
 
-  _checkIfInView() {
-    var selfOffset       = this.$().offset().top;
-    var scrollable       = this.get("_scrollable");
-    var scrollableBottom = scrollable.height() + scrollable.scrollTop();
+  _selfOffset() {
+    if(this.get('_customScrollableIsDefined')) {
```
Comment (davidgoli) [inline addon/components/infinity-loader.js]: use `if (` for consistency - a single space between keywords and open parens, to distinguish from functions, a la Crockford

File: addon/components/infinity-loader.js [hunk 12]
```diff
@@ -38,14 +39,32 @@ const InfinityLoaderComponent = Ember.Component.extend({
...
+      return this.$().position().top + this.get("_scrollable").scrollTop();
+    }
+    else {
+      return this.$().offset().top;
+    }
+  },
+
+  _bottomOfScrollableOffset() {
+    return this.get("_scrollable").height() + this.get("_scrollable").scrollTop();
```
Comment (davidgoli) [inline addon/components/infinity-loader.js]: prefer single quoted strings unless interpolating... this file isn't terribly consistent yet, but let's move in that direction

File: tests/acceptance/infinity-route-offset-trigger-test.js [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    infinityShouldNotBeReached(assert);
+    scrollTo(triggerOffset() - 100);
+  });
+  triggerEvent('ul', 'scroll');
+  andThen(() => {
+    shouldBeItemsOnTheList(assert, 25);
+    scrollTo(triggerOffset() + 100);
+  });
+  triggerEvent('ul', 'scroll');
+  andThen(() => {
```
Comment (davidgoli) [inline tests/acceptance/infinity-route-offset-trigger-test.js]: loosen these up with some individual newlines here please...

Comment (davidgoli) [inline tests/acceptance/infinity-route-offset-trigger-test.js]: for instance between `triggerEvent` and `andThen` and vice versa

**Comment by davidgoli** [comment]
Nice work

**Comment by davidgoli** [comment]
Just a few formatting comments but this looks solid... thanks for all the tests!

**Comment by hhff** [comment]
:+1: what @davidgoli - amazing work @mariuszzak  - very very keen to land this !

**Comment by mariuszzak** [comment]
Thanks a lot guys! I'm gonna improve code ASAP.

**Comment by mariuszzak** [comment]
Cosmetic improvements are finished. 

@davidgoli thanks again for comments! :100:

**Comment by hhff** [comment]
Nice!  I would say let's squash to a single commit and then we're ready to merge!

**Comment by mariuszzak** [comment]
Ok, done :)

**Pull request merged**